x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/frequentitemsets/TransactionStore.java

```diff
@@ -0,0 +1,769 @@
+ *
+ * NOTE: This class is not thread-safe!
+ *
+ * # Rough design
+ *
+ * ## Data:
+ *
+ * - transactions are split into items and items are stored together with their count
+ *    - every item has an id and a count
+ * - the transaction than gets rewritten into an array of items using the item ids, this dedups and compresses the storage
+ *    - every transaction is a bytebuffer containing item ids and a count
+ * - items and transactions are stored in a hashtable using ByteRefHash, which internally uses BigArrays
+ *    - the item store is basically a map: ByteRef -> item_id, the ByteRef key consists of the fieldname and its value,
+ *      the id is number (a count)
+ *    - the transaction store is a map: ByteRef -> transaction_id, the ByteRef consists of item_id's using a variable length
+ *      encoding (vlong)
+ * - the counts are stored as array, the index is {item_, transaction_}the id, the cell stores the count
+ * - the store can be serialized and send over the wire, serialization and deserialization uses variable length encodings to
+ *   make messages smaller
+ *
+ * ## Merging
+ *
+ * - a transaction store contains the items and transactions of 1 shard, this must be merged
+ * - transaction stores can be merged, this happens by rewriting and merging
+ *
+ * ## Pruning
+ *
+ * - prune removes all items that do not meet the support criteria and rewrites/drops transactions, this can significantly free memory
+ * - TODO: pruning is only used after merging all stores from all shards, it could be beneficial to prune already on the shard level
+ *
+ * ## Top Items/Transactions
+ *
+ * - top_{items, transactions} creates lists of {item_, transaction_}ids reverse sorted by count
+ *
+ */
+public final class TransactionStore implements Writeable, Releasable {
+
+    /**
+     * Container for holding item ids reverse sorted by count
+     *
+     * Only stores a list of ids, not the count
+     */
+    public static class TopItemIds implements Iterable<Long>, Releasable {
+
+        public class IdIterator implements Iterator<Long> {
+
+            private int currentIndex;
+
+            IdIterator(int startIndex) {
+                currentIndex = startIndex;
+            }
+
+            @Override
+            public boolean hasNext() {
+                return currentIndex < sortedItems.size();
+            }
+
+            @Override
+            public Long next() {
+                return sortedItems.get(currentIndex++);
+            }
+
+            @Override
+            public void remove() {
+                throw new UnsupportedOperationException();
+            }
+
+            public int getIndex() {
+                return currentIndex;
+            }
+
+            public void reset() {
+                currentIndex = 0;
+            }
+
+            public void reset(int startIndex) {
+                currentIndex = startIndex;
+            }
```

Comment:
I think many of these methods should be package private. Especially these `reset` methods.